### PR TITLE
Model key normalization

### PR DESCRIPTION
**Pull request by bizley**
Follow-up of the #18434 

This ensures that normalized key is a string. Also it introduces potential minor BC break (at the same time I don't know if anyone is using model keys in a way that can be broken here) but now it is behaving according to the original phpdoc (`return string` in both `normalizeModelKey` and `getModelKey`).

@rob006

### CODE REVIEW

**Comment by rob006** [comment]
I would double-check this change with more exotic AR implementations, like https://github.com/yiisoft/yii2-elasticsearch or https://github.com/yiisoft/yii2-sphinx - old implementation was suspiciously overcomplicated :P

**Comment by bizley** [comment]
So far ES is failing. I'll keep this thread updated with more news.

**Comment by bizley** [comment]
All looks good - or rather - all tests are green, so hopefully all is good.

File: framework/db/ActiveRelationTrait.php [hunk 1]
```diff
@@ -585,22 +585,23 @@ private function getModelKey($model, $attributes)
         if (count($key) > 1) {
             return serialize($key);
         }
-        $key = reset($key);
-        return is_scalar($key) ? $key : serialize($key);
+        return reset($key);
```
Comment (samdark) [inline framework/db/ActiveRelationTrait.php:588]: That was likely something to do with JSON columns that are both a single column in a key and isn't a scalar. I doubt anyone has these but still...

Comment (bizley) [reply framework/db/ActiveRelationTrait.php:588]: So do you think we should keep `normalizeModelKey` to return non-strings as well?

Comment (samdark) [reply framework/db/ActiveRelationTrait.php:588]: No, it won't work. Precisely, this piece will fail immediately:

```php
$key = $this->normalizeModelKey($key);
if (isset($buckets[$key])) {
    if ($this->indexBy !== null) {
        // if indexBy is set, array_merge will cause renumbering of numeric array
        foreach ($buckets[$key] as $bucketKey => $bucketValue) {
            $value[$bucketKey] = $bucketValue;
        }
    } else {
        $value = array_merge($value, $buckets[$key]);
    }
}
```

Comment (samdark) [reply framework/db/ActiveRelationTrait.php:588]: Yes, seems these changes are all correct.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Need a changelog line for it,.

**Comment by samdark** [comment]
👍

**Comment by zebraf1** [comment]
```php

class TestModel6 extends Model implements \JsonSerializable
{
    public $field5 = 'test5';
    public $field6 = 'test6';

    public function jsonSerialize()
    {
        return [
            'customField2' => $this->field5.'/'.$this->field6,
        ];
    }
}
```

Adding this to the SerializerTest:
```php

/**
     * @see https://github.com/yiisoft/yii2/issues/16334
     */
    public function testSerializeJsonSerializable()
    {
        $serializer = new Serializer();
        $model3 = new TestModel3();
        $model4 = new TestModel4();
        $model6 = new TestModel6();

        $this->assertEquals(['customField' => 'test3/test4'], $serializer->serialize($model3));
        $this->assertEquals(['customField2' => 'test5/test6'], $serializer->serialize($model4));
        $this->assertEquals(['customField2' => 'test5/test6'], $serializer->serialize($model6));
    }
```

Running tests:

There was 1 failure:

1) yiiunit\framework\rest\SerializerTest::testSerializeJsonSerializable
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    'customField2' => 'test5/test6'
+    'field5' => 'test5'
+    'field6' => 'test6'
 )

**Comment by zebraf1** [comment]
But I don't understand yet why it was necessary to change the order serialize() check for Arrayable and JsonSerializable. Was this causing an issue?

**Comment by bizley** [comment]
I don't see how this PR could break your code. Are you sure you don't think of maybe #18437 ?

**Comment by zebraf1** [comment]
Ah yes. I checked the commit that broke this was "Fix #18435: Change the check order whether an object is an implementation of `Arrayable` or `JsonSerializable` in `\yii\base\ArrayableTrait::toArray()` and `\yii\rest\Serializer::serialize()`"

But actually it did not reference it's own issue. Also changelog is incorrect and Spell6inder is referencing this issue instead.

**Pull request merged**